taichi/backends/vulkan/runtime.cpp

```diff
@@ -411,26 +407,46 @@ VkRuntime::KernelHandle VkRuntime::register_taichi_kernel(
 
 void VkRuntime::launch_kernel(KernelHandle handle, RuntimeContext *host_ctx) {
   auto *ti_kernel = ti_kernels_[handle.id_].get();
+
+  std::unique_ptr<DeviceAllocationGuard> ctx_buffer_host, ctx_buffer;
+
+  if (ti_kernel->get_ctx_buffer_size()) {
+    ctx_buffer = device_->allocate_memory_unique(
+        {ti_kernel->get_ctx_buffer_size(),
+           /*host_write=*/true, /*host_read=*/false,
+           /*export_sharing=*/false, AllocUsage::Storage});
+    ctx_buffer_host = device_->allocate_memory_unique(
+        {ti_kernel->get_ctx_buffer_size(),
+           /*host_write=*/false, /*host_read=*/true,
+           /*export_sharing=*/false, AllocUsage::Storage});
+  }
```

Comment:
r we creating new buffers every time we launch a kernel?

Reply:
yes

Reply:
.. that can't be good

Reply:
It is fine, it is actually faster than before

Reply:
be that as it may, we should still at least try to reuse some of these buffers. It is fast now because VMA is doing a superb job, but what if we want to unify the runtime code for different backends?

Reply:
We can't reuse them until we know for sure they are done. A better solution is a ring buffer, but in terms of memory usage it's the same.

Reply:
Also for DX11 and OpenGL this is going to be faster than a ring buffer because the drivers know for sure there isn't overlap and is free to use them async.

Reply:
So in the end only maybe Vulkan / DX12 / and Metal will benefit from a ring buffer, and all of these requires a robust memory allocator anyways.

Reply:
What about the overhead of allocating buffers on DX and GL? Are we confident that the performance gain from having things async outweighs these overheads?

Reply:
If it's buffer reuse it should be implemented in the allocator instead of the runtime

Reply:
Let's do the naive way for now, we can figure out a better way later